### DISCUSSION TITLE
AAGUID - attestationRootCertificate connection

### DISCUSSION THREAD

**Discussion by adam-siklosi** [webauthn4j/webauthn4j#941] [Q&A] [answered]
Hi @ynojima 

We have a more theroretical question about the verification of the attestation against the MDS `attestationRootCertificates`.

Webauthn4j uses the `AAGUID` to look up the certificate to validate against. https://github.com/webauthn4j/webauthn4j/blob/master/webauthn4j-core/src/main/java/com/webauthn4j/verifier/attestation/trustworthiness/certpath/DefaultCertPathTrustworthinessVerifier.java#L65

The specification is quite vague about this: https://www.w3.org/TR/webauthn-3/#reg-ceremony-assess-trust

> Otherwise, use the X.509 certificates returned as the attestation trust path from the verification procedure to verify that the attestation public key either correctly chains up to an acceptable root certificate, or is itself an acceptable certificate (i.e., it and the root certificate obtained in Step 22 may be the same).

`an acceptable certificate` is not very specific. :-)

The specification is also vague about the uniqueness of the AAGUID: https://www.w3.org/TR/webauthn-3/#sctn-authenticator-model

> Additionally, each authenticator has an AAGUID, which is a 128-bit identifier indicating the type (e.g. make and model) of the authenticator. The AAGUID MUST be chosen by the manufacturer to be identical across all substantially identical authenticators made by that manufacturer, and different (with high probability) from the AAGUIDs of all other types of authenticators. The AAGUID for a given type of authenticator SHOULD be randomly generated to ensure this. The Relying Party MAY use the AAGUID to infer certain properties of the authenticator, such as certification level and strength of key protection, using information from other sources.

`different (with high probability)` not exactly unique. :-) Probably here they are just protecting themselves against that practically they have no influence on manufacturers producing a non unique AAGUID.

The question is what was the basis for the `DefaultCertPathTrustworthinessVerifier` to use the AAGUID for looking up the `attestationRootCertificates`? Maybe I'm overlooking something in the spec.

Background: 
- We are looking into filtering authenticators based on AAGUID to allow only certain type in high security environments. The question come up what proves the connection of the AAGUID to the attestationRootCertificates. 
- A client could possibly submit a registration of an authenticator with an altered AAGUID matching an allowed AAGUID -  attestationRootCertificate pair. (low level attack / compromised browser)
- Though we are not sure if that is perhaps made impossible by some other mechanism on the client side, or perhaps this type of "attack" makes no sense at all.

Thanks,
Adam

**Comment by ynojima** [comment, marked as answer]
The AAGUID is an ID that the authenticator claims its identity. It is similar to a user agent string of a browser, as it can be easily spoofed and cannot be trusted on its own. 
To verify the trustworthiness of an authenticator, you need to have a root certificate for the attestation certificate issued to the trusted authenticator model. You then need to verify if the attestation certificate presented by the authenticator chains to the root certificate. webauthn4j does this on behalf of you.
If you want to trust multiple authenticator models, you need to find the root certificate corresponding to the authenticator. 
For this purpose, we made the `TrustAnchorRepository` interface which look up the trust anchor root certificate by AAGUID. 
A manipulated authenticator may return an altered AAGUID, but it cannot present an attestation certificate that chains to a trusted root certificate, so it won't be an issue.
